File: day02/day02.py
import sys


class Results:
    LOSE = 0
    TIE = 3
    WIN = 6
    Indexed = [LOSE, TIE, WIN]


# given both hands, give score of result
def get_result_score(theirs, yours):
    hand_differential = yours - theirs
    result = (hand_differential + 1) % 3
    return Results.Indexed[result]


# given opponent's hand & result, get score of your hand
def get_hand_score(op_hand, result):
    # win = 2 -> differential = +1, etc
    hand_differential = result - 1
    return ((op_hand - 1 + hand_differential) % 3) + 1


def get_numeric_round(str_round):
    [str_A, str_X] = str_round.split()
    int_A = ord(str_A) - ord("A") + 1
    int_X = ord(str_X) - ord("X") + 1
    return [int_A, int_X]


def calc_scores_1(rounds):
    score_sum = 0

    for round in rounds:
        [op_hand, your_hand] = get_numeric_round(round)
        result_score = get_result_score(op_hand, your_hand)
        score_sum += result_score + your_hand

    return score_sum


def calc_scores_2(rounds):
    score_sum = 0

    for round in rounds:
        [op_hand, result] = get_numeric_round(round)
        result_score = Results.Indexed[result - 1]
        hand_score = get_hand_score(op_hand, result - 1)
        score_sum += result_score + hand_score

    return score_sum


def load_input(path):
    lines = []
    with open(path, "r") as fp:
        for line in fp:
            lines.append(line.strip())
    return lines


def main():
    path = sys.argv[1] if len(sys.argv) > 1 else 'input.txt'
    input = load_input(path)

    result_1 = calc_scores_1(input)  # -> 13005
    result_2 = calc_scores_2(input)  # -> 11373

    print(result_1, result_2)


if __name__ == "__main__":
    main()

File: day02/day02.rb
# frozen_string_literal: true

R = { win: 6, tie: 3, lose: 0 }.freeze
HAND = { r: 1, p: 2, s: 3 }.freeze
R_INPUTS = [nil, R[:lose], R[:tie], R[:win]].freeze

WINNING_HANDS = {
  HAND[:r] => HAND[:p],
  HAND[:p] => HAND[:s],
  HAND[:s] => HAND[:r]
}.freeze

def get_hand(ltr)
  case ltr
  when 'A', 'X' then HAND[:r]
  when 'B', 'Y' then HAND[:p]
  when 'C', 'Z' then HAND[:s]
  end
end

def get_res(enemy, player)
  return R[:tie] if enemy == player

  return R[:win] if WINNING_HANDS[enemy] == player

  R[:lose]
end

def infer_hand(enemy, result)
  h = {
    R[:tie] => enemy,
    R[:win] => WINNING_HANDS[enemy],
    R[:lose] => WINNING_HANDS.key(enemy)
  }
  h[R_INPUTS[result]]
end

input = File.readlines(ARGV[0] || './input.txt', chomp: true)
            .map(&:split)
            .map { |match| match.map { |h| get_hand(h) } }

puts "A: #{input.map { |e, player| get_res(e, player) + player }.sum}"
puts "B: #{input.map { |e, r| infer_hand(e, r) + R_INPUTS[_r] }.sum}"
